src/uast.cc

```diff
@@ -137,95 +244,155 @@ void *UastIteratorNext(UastIterator *iter) {
+}
+
 Nodes *UastFilter(const Uast *ctx, void *node, const char *query) {
-  xmlDocPtr doc = nullptr;
-  xmlXPathContextPtr xpathCtx = nullptr;
-  xmlXPathObjectPtr xpathObj = nullptr;
-  bool ok = false;
-  int size = 0;
-  int realSize = 0;
-  int nodeIdx = 0;
-  xmlNodeSetPtr result = nullptr;
-  xmlNodePtr *results = nullptr;
-  xmlGenericErrorFunc handler;
-
-  Nodes *nodes;
-  try {
-    nodes = new Nodes();
-  } catch (std::bad_alloc) {
-    Error(nullptr, "Unable to get memory\n");
+  auto result = UastFilterTyped(ctx, node, query);
+
+  if (!checkResult(result, XPATHTYPE_NODESET, "UastFilter")) {
     return nullptr;
   }
-  doc = CreateDocument(ctx, node);
-  if (!doc) {
-    goto error1;
-  }
-  xpathCtx = static_cast<xmlXPathContextPtr>(xmlXPathNewContext(doc));
-  if (!xpathCtx) {
-    goto error2;
-  }
 
-  handler = (xmlGenericErrorFunc)Error;
-  initGenericErrorDefaultFunc(&handler);
+  return result.nodesVal;
+}
+
+int UastFilterBool(const Uast *ctx, void *node, const char *query) {
+  auto result = UastFilterTyped(ctx, node, query);
 
-  xpathObj = xmlXPathEvalExpression(BAD_CAST(query), xpathCtx);
-  if (!xpathObj) {
-    goto error3;
+  if (!checkResult(result, XPATHTYPE_BOOLEAN, "UastFilterBoolean")) {
+    return -1;
   }
 
-  // Get array of results
-  result = xpathObj->nodesetval;
-  if (!result) {
-    Error(nullptr, "Unable to get array of results\n");
-    goto error3;
-  }
-  results = result->nodeTab;
-  size = (result) ? result->nodeNr : 0;
-
-  // Sometimes libxml return null results for some invalid queries; do
-  // a first pass to discard them and get the real size for the result
-  for (int i = 0; i < size; i++) {
-    if (results[i] != nullptr && results[i]->_private != nullptr) {
-      ++realSize;
-    }
-  }
+  return result.boolVal;
+}
 
-  if (NodesSetSize(nodes, realSize) != 0) {
-    Error(nullptr, "Unable to set nodes size\n");
-    goto error3;
-  }
+double UastFilterNumber(const Uast *ctx, void *node, const char *query) {
+  auto result = UastFilterTyped(ctx, node, query);
 
-  // Populate array of results
-  for (int i = 0; i < size; i++) {
-    if (results[i] != nullptr && results[i]->_private != nullptr) {
-      nodes->results[nodeIdx++] = results[i]->_private;
-    }
+  if (!checkResult(result, XPATHTYPE_NUMBER, "UastFilterNumber")) {
+    return -1;
   }
-  ok = true;
 
-error3:
-  xmlXPathFreeObject(xpathObj);
-error2:
-  xmlXPathFreeContext(xpathCtx);
-error1:
-  xmlFreeDoc(doc);
+  return result.numberVal;
+}
+
+const char *UastFilterString(const Uast *ctx, void *node, const char *query) {
+  auto result = UastFilterTyped(ctx, node, query);
 
-  if (!ok) {
-    NodesFree(nodes);
+  if (!checkResult(result, XPATHTYPE_STRING, "UastFilterString")) {
     return nullptr;
   }
-  return nodes;
+
+  return result.stringVal;
 }
 
-char *LastError(void) {
-  return strndup(error_message, BUF_SIZE);
+UastTypedResult UastFilterTyped(const Uast *ctx, void *node, const char *query) {
```

Comment:
I don't think this function is very useful, quite the opposite. Different functions are calling this one to handle the shared behaviour but then there's a switch case for every kind of type, so no generic handling at all...

Reply:
Yes, initially I only added this to the public API (without the typed functions) I left it that way in case we decided to use it instead of the typed ones, I'll refactor.